langchain/src/schema/retriever.ts

```diff
@@ -0,0 +1,71 @@
+import {
+  BaseCallbackConfig,
+  CallbackManager,
+  CallbackManagerForRetrieverRun,
+  Callbacks,
+  parseCallbackConfigArg,
+} from "../callbacks/manager.js";
+import { Document } from "../document.js";
+import { Serializable } from "../load/serializable.js";
+
+/**
+ * Base Index class. All indexes should extend this class.
+ */
+
+export interface BaseRetrieverInput {
+  callbacks?: Callbacks;
+  tags?: string[];
+  metadata?: Record<string, unknown>;
+  verbose?: boolean;
+}
+
+export abstract class BaseRetriever extends Serializable {
+  callbacks?: Callbacks;
+
+  tags?: string[];
+
+  metadata?: Record<string, unknown>;
+
+  verbose?: boolean;
+
+  constructor(fields?: BaseRetrieverInput) {
+    super(fields);
+    this.callbacks = fields?.callbacks;
+    this.tags = fields?.tags ?? [];
+    this.metadata = fields?.metadata ?? {};
+    this.verbose = fields?.verbose ?? false;
+  }
+
+  abstract _getRelevantDocuments(
```

Comment:
For folks who might be using custom retrievers, might be best to not have this as an `abstract` method and just throw a `NotImplemented` error or something similar? And add a note to make it abstract on the 0.1.0/1.0.0 release.